draft-wang-ppm-dap-taskprov.md

```diff
@@ -222,144 +225,212 @@ struct {
 
-The client should know whether `task-prov` extension will be used and all
-parameters required for `TaskConfig` prior to constructing the extension
-body, either out-of-band from aggregators, or from information already saved on
-client.
+> OPEN ISSUE: Consider using a randomness extractor, such has `HKDF-Extract()`,
+> so that we can treat the `task_id` as pseudorandom in cases where
+> `task_config` has high min-entropy
 
-To offer the `task-prov` extension, the client adds the `TaskConfig` structure
-it received from the task author in the extensions field of its `Report`. It
-computes the task ID as described in {{construct-task-id}}.
+> TODO: Add RFC reference for SHA-256.
 
-## Construct task ID {#construct-task-id}
+## Deriving the VDAF Verification Key {#vdaf-verify-key}
 
-For `task-prov` extension, a DAP task is not created before distributing task
-configuration to clients. Therefore, clients, aggregators and collector
-construct the DAP task ID prior to uploading. A DAP task ID is computed as
-follows:
+> TODO
 
-~~~
-task_id = SHA-256(task_config)
-~~~
+## Configuring a Task {#provisioning-a-task}
 
-# Provisioning a task {#provisioning-a-task}
+Prior to participating in a task, each protocol participant must determine if
+the `TaskConfig` disseminated by the Author can be configured. The participant
+is said to "opt in" to the task if the derived task ID (see
+{{construct-task-id}}) corresponds to an already configured task or the task ID
+is unrecognized and therefore corresponds to a new task.
 
-Upon receiving the payload containing extension, leader and helpers perform
-similar steps to provision a task.
+A protocol participant MAY "opt out" of a task if:
 
-If aggregator supports `task-prov` extension, it should first check if the
-task ID already exists, if so the aggregator continues to the rest of the flow
-being processed. Note if the existing tasks's configuration is different from
-the one in extension, HPKE decryption will fail due to mismatched AAD.
+1. The derived task ID corresponds to an already configured task, but the task
+   configuration disseminated by the Author does not match the existing
+   configuration.
 
-If the task ID has not been seen before, aggregator should read and decode
-`extension_data` with the `TaskConfig` schema. If the decoding failed, it MUST
-abort the sub protocol with error "unrecognizedMessage".
+1. The VDAF, DP, or query configuration is deemed insufficient for privacy.
 
-If the decoding succeeds, aggregator creates a new task using the task ID
-from the decoded extension, and save task configuration with the newly created
-task. In particular, aggregator should deserialize `vdaf_config` corresponding
-to `vdaf_type`, and pass the relevant parameters to the VDAF initializer. At
-this point, the task provision step has completed.
+1. A secure connection to one or both of the Aggregator endpoints could not be
+   established.
 
-# Leader Behavior
+1. The task lifetime is too long.
```

Comment:
The task_expiration is what I'm referring to. The behavior described here is: "A party MAY opt out if it thinks the task lives too long".

Reply:
My understanding is that task_expiration is basically a timestamp. If this is so, then I don't think this is ambiguous, since the MAY applies to expiration dates in the future, whereas the MUST applies to expiration dates in the past.

However there is another problem here. The goal of this MAY is to allow a party to opt-out if the task is going too long. There are two reasons why it might want to do this:
1. Sanity checking: We want to make sure tasks don't live for 100s of years.
2. For a given deployment of DAP, the parties might make an agreement that they will keep around metadata for anti-replay for as long as the task exists. In this case it would be useful to check that the task exists for, say, no longer than a week so that they don't have to keep around the metadata for longer than needed.

In either case, what's missing is the task's "start time". This is important because the Author might have created the task minutes or even hours before the Aggregators (in particular the Helper) sees the task for the first time.

If you agree this is needed, I'll add it to this PR.

Reply:
for case1, isn't the task_expiration sufficient? I'm not convinced a start time is the solution but I'm curious how this start time looks like, perhaps we can keep the current PR text and create another PR to address this?

Reply:
> for case1, isn't the task_expiration sufficient?

Deciding whether a task has a lifetime of at least 200 years, I would subtract task_start from task_expiration and see if the duration is > 200 years. We can't do this correctly without knowing task_start.

> I'm not convinced a start time is the solution but I'm curious how this start time looks like, perhaps we can keep the current PR text and create another PR to address this?

Sure, makes sense! Raised https://github.com/wangshan/draft-wang-ppm-dap-taskprov/issues/12